torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -3496,62 +3500,91 @@ def full_optim_state_dict(
+                    "in `full_optim_state_dict()`."
+                )
+            group = next(iter(process_groups))
         rank = dist.get_rank(group)
         to_save = not rank0_only or rank == 0
         full_osd: Dict = {"state": {}, "param_groups": []} if to_save else {}
-        full_osd_state = full_osd["state"] if to_save else None  # alias
-
-        # Handle the "state" part of the optimizer state dict
-        param_to_unflat_param_names = _get_param_to_unflat_param_names(model)
-        flat_param_id_to_param = _get_param_id_to_param(model, optim_input)
-        flat_param_to_fsdp_module = _get_flat_param_to_fsdp_module(model)
-        for flat_param_id, param in enumerate(flat_param_id_to_param):  # type: ignore[assignment]
+        full_osd_state = full_osd["state"] if to_save else None
+
+        param_to_unflat_param_names: Dict[torch.nn.Parameter, List[str]] = \
+            _get_param_to_unflat_param_names(model)
+        flat_param_id_to_param: List[torch.nn.Parameter] = \
+            _get_param_id_to_param(model, optim_input)
+        optim_state_key_to_flat_param_id: Dict[_OptimStateKey, int] = {}  # local
+        r0_flat_param_id_to_optim_state_key: Dict[int, _OptimStateKey] = OrderedDict()  # rank 0
+        for flat_param_id, param in enumerate(flat_param_id_to_param):
             # Do not include parameters without state to avoid empty mappings
+            # just like in normal `torch.optim.Optimizer.state_dict()`
             if flat_param_id not in osd_state:
                 continue
-            assert param in param_to_unflat_param_names, \
-                "Check the `param_to_unflat_params` construction\n" \
-                f"param: {param}"
-            unflat_param_names = param_to_unflat_param_names[param]
-            # For FSDP parameters, we need to unflatten
-            if isinstance(param, FlatParameter):
-                assert param in flat_param_to_fsdp_module, \
-                    "Check the `flat_param_to_fsdp_module` construction\n" \
-                    f"param: {param}"
-                unflat_state = _unflatten_optim_state(
-                    flat_param_to_fsdp_module[param], param,
-                    osd_state[flat_param_id], to_save,
-                )
+            optim_state_key = _OptimStateKey(
+                unflat_param_names=tuple(param_to_unflat_param_names[param]),
+                is_flat_param=isinstance(param, FlatParameter),
+            )
+            if rank == 0:
+                r0_flat_param_id_to_optim_state_key[flat_param_id] = optim_state_key
+            optim_state_key_to_flat_param_id[optim_state_key] = flat_param_id
+        obj_list = [r0_flat_param_id_to_optim_state_key] if rank == 0 else [None]
+        dist.broadcast_object_list(obj_list, src=0, group=group)
+        r0_flat_param_id_to_optim_state_key = obj_list[0]
+
+        # Ensure that all ranks have at least the optimizer states needed by
```

Comment:
I think this validation is important to prevent hangs. All ranks should agree that they can proceed into the upcoming `for` loop over rank 0's `r0_optim_state_key`s without any issues before doing so. Otherwise, one rank might error, causing the others to hang.